compiler/rustc_resolve/src/macros.rs

```diff
@@ -694,7 +694,25 @@ impl<'a> Resolver<'a> {
                     check_consistency(self, &path, path_span, kind, initial_res, res)
                 }
                 path_res @ PathResult::NonModule(..) | path_res @ PathResult::Failed { .. } => {
+                    let mut suggestion = None;
                     let (span, label) = if let PathResult::Failed { span, label, .. } = path_res {
+                        // try to suggest if it's not a macro, maybe a function
+                        if let PathResult::NonModule(partial_res) =  self.resolve_path(
+                            &path,
+                            Some(ValueNS),
+                            &parent_scope,
+                            Some(Finalize::new(ast::CRATE_NODE_ID, path_span)),
+                            None,
+                        ) && partial_res.unresolved_segments() == 0 {
+                            let sm = self.session.source_map();
+                            let span = sm.span_extend_while(span, |c| c == '!').unwrap_or(span);
```

Comment:
Can you make the suggestion just remove the `!`, and not replace `macro!` with `macro`? I think you can use something like `SourceMap::next_point`

Reply:
I have a question for the `next_point`:


For the debug code:

```rust
 let exclamation_span = sm.next_point(span);
                            debug!("debug-now span: {:?}", span);
                            debug!("debug-now exclamation_span: {:?}", exclamation_span);
                            debug!("debug-now exclamation_span source: {:?}", sm.span_to_snippet(exclamation_span));
                            debug!("debug-now exclamation_span next : {:?}", sm.next_point(exclamation_span));
```

Printed out this log:
```console
DEBUG rustc_resolve::macros debug-now span: src/test/ui/suggestions/issue-103112.rs:2:19: 2:24 (#0)
DEBUG rustc_resolve::macros debug-now exclamation_span: src/test/ui/suggestions/issue-103112.rs:2:24: 2:24 (#0)
DEBUG rustc_resolve::macros debug-now exclamation_span source: Ok("")
DEBUG rustc_resolve::macros debug-now exclamation_span next : src/test/ui/suggestions/issue-103112.rs:2:24: 2:25 (#0)
```

Why `span` (with lo and hi 2:19: 2:24) 's `next_point` is not (2:25 2:25), it's (2:24 2:24) instead? seems counterintuitive.

Reply:
There is a bug at here?

https://github.com/rust-lang/rust/blob/194140bef501ad3acb00d57c20fb80ee34aa1d3b/compiler/rustc_span/src/source_map.rs#L869

```diff
-        let end_of_next_point = BytePos(cmp::max(sp.lo().0 + 1, end_of_next_point));
+        let end_of_next_point = BytePos(cmp::max(sp.hi().0 + 1, end_of_next_point));
```

Reply:
Because of this bug, we always want to all two times of `next_point`， 

For example:
https://github.com/rust-lang/rust/blob/9b0a099dfc9a97ecd10adb319396c731c4b2d169/compiler/rustc_expand/src/expand.rs#L940

https://github.com/rust-lang/rust/blob/98a5ac269cffada469753ad2416717e251863f9a/compiler/rustc_resolve/src/late/diagnostics.rs#L1733

@estebank how do you think?

Reply:
Add another PR to fix it: https://github.com/rust-lang/rust/pull/103185

Reply:
@chenyukang thanks for fixing the bug with `next_point`!